coll: remove TSP namespace swapping

hzhou opened the pull request:
## Pull Request Description

The idea of TSP namespace is to have the same TSP algorithm realized with different transport. However, currently, we only have the gentran transport, and it is not clear whether we will ever develop another transport (versus enhancing the gentran transport), and even when we do, it is not clear whether there is a benefit to provide multiple realizations at the runtime. Likely we will simply pick a better performing transport and remove another.

The current mechanism of support this TSP namespace swap for the purpose of providing multiple realizations of a single TSP algorithm is complicated and difficult to maintain. Since there is little foreseeable benefit, this PR removes this mechanism.

`stubtran` is removed.

The TSP interface is defined in `tsp_impl.h`. All gentran internals are defined with `MPII_Genutil_` prefix and not visible outside the gentran interface.

Fixes #5265

[skip warnings]

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [x] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

hzhou commented:
test:mpich/ch3/tcp
test:mpich/ch4/ofi

hzhou commented:
test:mpich/ch4/ofi

zhenggb72 commented:
It seems the PR kept the name "TSP" but disabled transport swapping. I don't think it worth introducing another name TSP now that TSP == GenTran. I think it would be more clear if functions are named directly with Gentran instead of TSP, e.g. MPIR_Gentran_Ibcast_intra_tree instead of MPIR_TSP_Ibcast_intra_tree.

hzhou commented:
> 
> 
> It seems the PR kept the name "TSP" but disabled transport swapping. I don't think it worth introducing another name TSP now that TSP == GenTran. I think it would be more clear if functions are named directly with Gentran instead of TSP, e.g. MPIR_Gentran_Ibcast_intra_tree instead of MPIR_TSP_Ibcast_intra_tree.

There are two reasons. 1. Less churn of the code since TSP is used everywhere (except the algorithm cvar names). 2. Let's say we later want to test another transport mechanism, the algorithms will stay the same but the transport can be swapped out with a set of macros. 

So I am more leaning toward rename the algorithm names, which I understand it is also undesirable churn. I think it may do for now. We always can do the rename later once we reach some consensus, or when we clarify the future plan (whether we will experiment another transport).

zhenggb72 commented:
If we have any future plan to experiment another transport (which I doubt it), then we should actually keep TSP as it was before this PR.
But as it is now TSP == Gentran. I still prefer either renaming all Gentran to TSP, or all TSP to Gentran, but not both, which may be confusing to other people.

hzhou commented:
test:mpich/ch3/tcp
test:mpich/ch4/ofi

hzhou commented:
> 
> 
> If we have any future plan to experiment another transport (which I doubt it), then we should actually keep TSP as it was before this PR.
> But as it is now TSP == Gentran. I still prefer either renaming all Gentran to TSP, or all TSP to Gentran, but not both, which may be confusing to other people.

Added last 5 commits to simplify the TSP interface and rename everything into TSP names except Gentran internal utilities.

raffenet reviewed (approved):
👍 for simplification and code removal in this PR.

Long-term: I wonder, is TSP just an alternative sched interface? If so, what the pros/cons over the normal sched interface? Can the 2 be merged?

hzhou commented:
> Long-term: I wonder, is TSP just an alternative sched interface? If so, what the pros/cons over the normal sched interface? Can the 2 be merged?

From what I gathered, TSP design tracks operation dependency, so it may be possible to implement some pipelining optimization. The normal sched interface can be implemented with the TSP interface, but we lose the simplicity and flexibility. Our nonblocking context_id algorithm is very hacky, so it may not be possible with the TSP interface.

zhenggb72 commented on src/mpi/coll/transports/gentran/tsp_gentran.c:
```diff
@@ -92,25 +92,6 @@ void MPII_Genutil_sched_free(MPII_Genutil_sched_t * sched)
-    return type_id;
-}
-
 int MPII_Genutil_sched_generic(int type_id, void *data,
```
can we undo this commit?
The new_type interface is actually used by us. This interface is important so that we can add new node type in the schedule tree.

hzhou replied:
I see. So we have to expose the internal gentran data structure?

zhenggb72 replied:
I don't see how it really exposes the internal data structure. But I agree it is specific to Gentran.

hzhou replied:
The function pointers would need the vertex data struct definitions, right?

zhenggb72 replied:
You are probably right. anyway what is the issue with that? It is just part of the gentran API.

hzhou replied:
Oh, I was trying to make the TSP interface abstract. Abstract interfaces are easier to maintain and extend. Since there is code breaking that abstraction, I guess we'll just deal with it.

I am a bit curious how did it happen. Isn't the whole design point originally was to keep TSP interface abstract so we could swap the gentran with alternate transport?

zhenggb72 replied:
Swappable TSP interface probably never worked. The latest idea is to make gentran more general and extendable with this new interface so that it works for new algorithms.

hzhou replied:
ok

hzhou replied:
@zhenggb72 In the prototypes of callback functions:
https://github.com/zhenggb72/mpich/blob/af753cae3b43e5e12a4030f6841328b1d238c917/src/mpi/coll/transports/gentran/tsp_gentran_types.h#L126-L128
Is it sufficient to pass in the `void *data` instead of `MPII_Genutil_vtx_t * vtxp`?

hzhou replied:
@zhenggb72 The mentioned commit has been updated in https://github.com/pmodels/mpich/pull/5313/commits/f209fa76cf036aa244368c15322b2bb47f32431d. Please review.

hzhou commented:
NOTE: the PR that added `MPII_Genutil_sched_new_type` is https://github.com/pmodels/mpich/pull/3872

hzhou commented:
@raffenet There some significant additions/changes since last time you reviewed this PR. Mainly --
* `MPIR_TSP_sched_t` has been made opaque (as `void *`). This is very similar to `MPIR_Sched_t`.
* All `Genutil` names now are kept internal to the gentran transport. Outside the transport, we only use `TSP` namespace. All previous *gentran* algorithms are renamed into *tsp* algorithms.
* I made sure each commit can compile on its own.

I will clean up the warnings and run a few additional tests. Meanwhile, would you like to give it another review?

hzhou commented:
test:mpich/ch4/most
test:mpich/ch3/most

hzhou commented:
test:mpich/ch4/most
